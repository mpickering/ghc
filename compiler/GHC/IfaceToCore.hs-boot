module GHC.IfaceToCore where

import GhcPrelude
import GHC.Iface.Syntax ( IfaceDecl, IfaceClsInst, IfaceFamInst, IfaceRule
                        , IfaceAnnotation, IfaceCompleteMatch, IfaceExpr, IfaceType )
import GHC.Core.TyCo.Rep   ( TyThing, Type )
import TcRnTypes           ( IfL )
import GHC.Core.InstEnv    ( ClsInst )
import GHC.Core.FamInstEnv ( FamInst )
import GHC.Core         ( CoreRule, CoreExpr )
import GHC.Driver.Types ( CompleteMatch )
import Annotations      ( Annotation )
import Name (Name)

tcIfaceDecl         :: Bool -> IfaceDecl -> IfL TyThing
tcIfaceRules        :: Bool -> [IfaceRule] -> IfL [CoreRule]
tcIfaceInst         :: IfaceClsInst -> IfL ClsInst
tcIfaceFamInst      :: IfaceFamInst -> IfL FamInst
tcIfaceAnnotations  :: [IfaceAnnotation] -> IfL [Annotation]
tcIfaceCompleteSigs :: [IfaceCompleteMatch] -> IfL [CompleteMatch]
tcIfaceGlobal :: Name -> IfL TyThing
tcIfaceExpr :: IfaceExpr -> IfL CoreExpr
tcIfaceType :: IfaceType -> IfL Type
