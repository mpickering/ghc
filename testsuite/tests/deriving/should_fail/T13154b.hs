{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE UnboxedTuples #-}
module T13154b where

import GHC.Exts

-- Test some nonsense configurations

class Foo (a :: TYPE ('TupleRep '[]))
deriving stock   instance Foo a
deriving stock   instance Foo (##)
deriving newtype instance Foo a
deriving newtype instance Foo (##)

class C
deriving stock   instance C
deriving newtype instance C
