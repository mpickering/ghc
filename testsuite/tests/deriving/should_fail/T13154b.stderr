
T13154b.hs:15:1: error:
    • Can't make a derived instance of ‘Foo a’ with the stock strategy:
        ‘Foo’ is not a stock derivable class (Eq, Show, etc.)
    • In the stand-alone deriving instance for ‘Foo a’

T13154b.hs:16:1: error:
    • Can't make a derived instance of
        ‘Foo (# #)’ with the stock strategy:
        ‘Foo’ is not a stock derivable class (Eq, Show, etc.)
    • In the stand-alone deriving instance for ‘Foo (# #)’

T13154b.hs:17:1: error:
    • Can't make a derived instance of
        ‘Foo a’ with the newtype strategy:
        GeneralizedNewtypeDeriving cannot be used on non-newtypes
    • In the stand-alone deriving instance for ‘Foo a’

T13154b.hs:18:1: error:
    • Can't make a derived instance of
        ‘Foo (# #)’ with the newtype strategy:
        GeneralizedNewtypeDeriving cannot be used on non-newtypes
    • In the stand-alone deriving instance for ‘Foo (# #)’

T13154b.hs:21:1: error:
    • Can't make a derived instance of ‘C’ with the stock strategy:
        ‘C’ is not a stock derivable class (Eq, Show, etc.)
    • In the stand-alone deriving instance for ‘C’

T13154b.hs:22:1: error:
    • Can't make a derived instance of ‘C’ with the newtype strategy:
        GeneralizedNewtypeDeriving cannot be used on non-newtypes
    • In the stand-alone deriving instance for ‘C’
